### ISSUE TITLE
A sleep of 1 second returns immediately

### ISSUE THREAD

**Issue by daurnimator** [wahern/cqueues#75]
I was messing around, and with the current HEAD, you can't sleep for 1 second....
This is really weird:
```
$ time lua -e 'cqueues=require"cqueues"; print("POLL", cqueues.poll(1)) '
POLL

real	0m0.005s
user	0m0.003s
sys	0m0.000s
```
```
$ time lua -e 'cqueues=require"cqueues"; print("POLL", cqueues.poll(2)) '
POLL

real	0m1.004s
user	0m0.000s
sys	0m0.000s
```
```
$ time lua -e 'cqueues=require"cqueues"; print("POLL", cqueues.poll(1.1)) '
POLL	1.1

real	0m1.105s
user	0m0.000s
sys	0m0.000s
```
```
$ time lua -e 'cqueues=require"cqueues"; print("POLL", cqueues.poll(0.9)) '
POLL	0.9

real	0m0.905s
user	0m0.000s
sys	0m0.003s
```

Just noticed that the 2 second sleep is sleeping for only 1 second too...
==> Also the faulty numbers don't seem to return the arguments from `poll`

Note that it only seems to be an issue when using `cqueues.poll` outside a cqueue.

**Comment by daurnimator** [comment]
The issue seems to be in the double => timestamp conversion.

Patch idea:
```diff
diff --git a/src/cqueues.c b/src/cqueues.c
index 3bad8fe..65f5f87 100644
--- a/src/cqueues.c
+++ b/src/cqueues.c
@@ -197,7 +197,7 @@ static inline int f2ms(const double f) {
 			return INT_MAX;
 		}
 
-		return ((int)f * 1000) + ((int)ceil(f * 1000.0) % 1000);
+		return (int)ceil(f * 1000.0);
 	case FP_SUBNORMAL:
 		return 1;
 	case FP_ZERO:
@@ -219,8 +219,12 @@ static inline struct timespec *f2ts_(struct timespec *ts, const double f) {
 			ts->tv_nsec = 0;
 		} else {
 			ts->tv_sec = (time_t)f;
+			unsigned long tmp = ceil(f * 1000000000.0);
 			/* SunPRO chokes on modulo here unless unsigned. */
-			ts->tv_nsec = (unsigned long)ceil(f * 1000000000.0) % 1000000000UL;
+			ts->tv_nsec = tmp % 1000000000UL;
+			if (ts->tv_nsec == 0 && tmp != 0) {
+				ts->tv_sec++;
+			}
 		}
 
 		return ts;
```

I'm not sure why `f2ms` has that complicated sum anyway... a plain `ceil` should work, no?

The `f2ts` code in the patch is untested...
But looking at it, even the existing code seems wrong. `ceil(f * 1000000000.0)` is only rarely going to fit in an `unsigned long`...

**Comment by daurnimator** [comment]
This might be better:

```diff
diff --git a/src/cqueues.c b/src/cqueues.c
index 3bad8fe..c107e1d 100644
--- a/src/cqueues.c
+++ b/src/cqueues.c
@@ -30,7 +30,7 @@
 #include <string.h>	/* memset(3) */
 #include <signal.h>	/* sigprocmask(2) pthread_sigmask(3) */
 #include <time.h>	/* struct timespec clock_gettime(3) */
-#include <math.h>	/* FP_* NAN fmax(3) fpclassify(3) isfinite(3) signbit(3) islessequal(3) isgreater(3) ceil(3) */
+#include <math.h>	/* FP_* NAN fmax(3) fpclassify(3) isfinite(3) signbit(3) islessequal(3) isgreater(3) ceil(3) floor(3) */
 #include <errno.h>	/* errno */
 #include <assert.h>	/* assert */
 
@@ -197,7 +197,7 @@ static inline int f2ms(const double f) {
 			return INT_MAX;
 		}
 
-		return ((int)f * 1000) + ((int)ceil(f * 1000.0) % 1000);
+		return (int)ceil(f * 1000.0);
 	case FP_SUBNORMAL:
 		return 1;
 	case FP_ZERO:
@@ -218,9 +218,13 @@ static inline struct timespec *f2ts_(struct timespec *ts, const double f) {
 			ts->tv_sec = (time_t)INT_MAX;
 			ts->tv_nsec = 0;
 		} else {
-			ts->tv_sec = (time_t)f;
-			/* SunPRO chokes on modulo here unless unsigned. */
-			ts->tv_nsec = (unsigned long)ceil(f * 1000000000.0) % 1000000000UL;
+			double sec = floor(f);
+			ts->tv_sec = (time_t)sec;
+			ts->tv_nsec = ceil((f-sec) * 1000000000.0);;
+			if (ts->tv_nsec >= 1000000000.0) {
+				ts->tv_nsec -= 1000000000.0;
+				ts->tv_sec++;
+			}
 		}
 
 		return ts;
```

The `if` has to be there for the case of `0.9999999999`: it gets rounded down by `floor`, but up by `ceil`

**Comment by daurnimator** [comment]
Fixed with a379837
